### ISSUE TITLE
References break when the ID values are compound IDs

### ISSUE THREAD

**Issue by H4ml3t** [mongodb/morphia#675]
Hi all, first, let me tell you what I think of morphia: WOW!
I'm a new user, so I'm not sure this is really an issue or I'm doing something wrong.
I have this two classes to store (you can find them below): one is representing an Execution e the other one a Message. One Execution contains more Messages and one Message belongs just to one Execution. I want the key of an Execution to be like this: *(username,nodename,time)*.

In the following snippet of code everything work BUT last *get* operation (OP5):
```java
Datastore ds = ...
Date currentDate = new Date();
ExecutionKey key = new ExecutionKey("usernameTest", "namenodeTest", currentDate);
ExecutionInterface execution = new ExecutionMorphiaImpl(key,"0.0.0-test");
ds.save(execution); // OP1
String result1 = ds.get(ExecutionMorphiaImpl.class, 
	new ExecutionKey("usernameTest","namenodeTest",currentDate)).toString(); // OP2
MessageInterface m1 = new MessageMorphiaImpl(execution,"hdfs operation","info","hadoop fs -ls thispath",null);
MessageInterface m2 = new MessageMorphiaImpl(execution,"spark job","info","spark-submit this.jar ...",null);
MessageInterface m3 = new MessageMorphiaImpl(execution,"spark job","warn","submitting this.jar ...","exit code 255");
ds.save(m1,m2,m3); // OP3
execution.getMessageList().add(m1);
execution.getMessageList().add(m2);
execution.getMessageList().add(m3);
ds.save(execution); // OP4
String result2 = ds.get(ExecutionMorphiaImpl.class, 
	new ExecutionKey("usernameTest","namenodeTest",currentDate)).toString(); // OP5
```

I can actually find the result inside mongodb for the operations OP1, OP2, OP3 and OP4.
But with OP5 I got this error:
```
Nov 25, 2014 1:03:30 AM org.apache.catalina.core.StandardWrapperValve invoke
SEVERE: Servlet.service() for servlet [Jersey REST Service] in context with path [] threw exception
org.mongodb.morphia.mapping.MappingException: Could not map com.morphiatest.morphia.model.ExecutionMorphiaImpl with ID: { "username" : "usernameTest" , "nodename" : "namenodeTest" , "time" : { "$date" : "2014-11-25T00:03:30.764Z"}}
        at org.mongodb.morphia.mapping.Mapper.fromDb(Mapper.java:587)
        at org.mongodb.morphia.mapping.Mapper.fromDBObject(Mapper.java:292)
        at org.mongodb.morphia.query.MorphiaIterator.convertItem(MorphiaIterator.java:79)
        at org.mongodb.morphia.query.MorphiaIterator.processItem(MorphiaIterator.java:65)
        at org.mongodb.morphia.query.MorphiaIterator.next(MorphiaIterator.java:60)
        at org.mongodb.morphia.query.QueryImpl.get(QueryImpl.java:395)
        at org.mongodb.morphia.DatastoreImpl.get(DatastoreImpl.java:664)
        at com.morphiatest.servlet.test.morphiaTest(test.java:61)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:483)
        at com.sun.jersey.spi.container.JavaMethodInvokerFactory$1.invoke(JavaMethodInvokerFactory.java:60)
        at com.sun.jersey.server.impl.model.method.dispatch.AbstractResourceMethodDispatchProvider$TypeOutInvoker._dispatch(AbstractResourceMethodDispatchProvider.java:185)
        at com.sun.jersey.server.impl.model.method.dispatch.ResourceJavaMethodDispatcher.dispatch(ResourceJavaMethodDispatcher.java:75)
        at com.sun.jersey.server.impl.uri.rules.HttpMethodRule.accept(HttpMethodRule.java:288)
        at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147)
        at com.sun.jersey.server.impl.uri.rules.ResourceClassRule.accept(ResourceClassRule.java:108)
        at com.sun.jersey.server.impl.uri.rules.RightHandPathRule.accept(RightHandPathRule.java:147)
        at com.sun.jersey.server.impl.uri.rules.RootResourceClassesRule.accept(RootResourceClassesRule.java:84)
        at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1469)
        at com.sun.jersey.server.impl.application.WebApplicationImpl._handleRequest(WebApplicationImpl.java:1400)
        at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1349)
        at com.sun.jersey.server.impl.application.WebApplicationImpl.handleRequest(WebApplicationImpl.java:1339)
        at com.sun.jersey.spi.container.servlet.WebComponent.service(WebComponent.java:416)
        at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:537)
        at com.sun.jersey.spi.container.servlet.ServletContainer.service(ServletContainer.java:699)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:728)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:305)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:210)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:222)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:123)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:472)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:171)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:99)
        at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:931)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:118)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:407)
        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1004)
        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:589)
        at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:310)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: org.mongodb.morphia.mapping.MappingException: The reference({ "$ref" : "Message", "$id" : "5473c752e4b0f43ae5250707" }) could not be fetched for com.morphiatest.morphia.model.ExecutionMorphiaImpl.messageList
        at org.mongodb.morphia.mapping.ReferenceMapper.resolveObject(ReferenceMapper.java:331)
        at org.mongodb.morphia.mapping.ReferenceMapper$1.eval(ReferenceMapper.java:260)
        at org.mongodb.morphia.utils.IterHelper.loopOrSingle(IterHelper.java:70)
        at org.mongodb.morphia.mapping.ReferenceMapper.readCollection(ReferenceMapper.java:257)
        at org.mongodb.morphia.mapping.ReferenceMapper.fromDBObject(ReferenceMapper.java:174)
        at org.mongodb.morphia.mapping.Mapper.readMappedField(Mapper.java:605)
```

**Class ExecutionMorphiaImpl**
```java
@Entity("Execution")
public class ExecutionMorphiaImpl implements ExecutionInterface {

	@Embedded
	public static class ExecutionKey implements Serializable {
		private static final long serialVersionUID = 1L;
		private String username;
		private String nodename;
		private Date time;
		public ExecutionKey() {}
		public ExecutionKey(String username, String nodename, Date time) {
			this.username = username;
			this.nodename = nodename;
			this.time = time;
		}
		@Override public int hashCode() { ... }
		@Override public boolean equals(Object obj) { ... }
		@Override public String toString() { ... }

	}

	@Id
	private ExecutionKey key;
	@Indexed(unique=true)
	private ObjectId id; // an Id for the rest api
	@Indexed
	private String username;
	@Reference
	private ArrayList<MessageInterface> messageList;
	@Indexed
	private String version;
	public ExecutionMorphiaImpl() {}
	public ExecutionMorphiaImpl(ExecutionKey key, String version) {
		this.key = key;
		this.id = new ObjectId(key.time); //Problem #2, is there a way to automatize this?
		this.username = key.username;
		this.nodename = key.nodename;
		this.time = key.time;
		this.version = version;
	}
```

**class MessageMorphiaImpl**
```java
@Entity("Message")
public class MessageMorphiaImpl implements MessageInterface {
	@Reference
	private ExecutionInterface execution;
	@Indexed
	private String type;
	@Indexed
	private String level;
	private String content;
	private String result;
	@Id
	private ObjectId id;

	public MessageMorphiaImpl() { }
	public MessageMorphiaImpl(ExecutionInterface execution, String type, String level, String content, String result) {
		this.type = type;
		this.level = level;
		this.content = content;
		this.result = result;
		this.execution = execution;
	}
```

**Comment by evanchooly** [comment]
OK.  Two things:

1.  This is less of a mapping problem and more of workflow problem.  You need to save any referenced objects/documents first so that they have ID values to store in the DBRef in the database.  That's what the error is actually saying.
1.  This particular validation has been removed because it was a pointless performance hit.

Please try again with the latest version and with saving your referenced objects first.

**Comment by evanchooly** [comment]
To clear the slate for 1.0, I'm going to bump this to post-1.0.  I'll keep it open for a while until I hear back from you or I'll just close it eventually  either way.  I'm reasonably sure there's no bug here now.

**Comment by H4ml3t** [comment]
>You need to save any referenced objects/documents first so that they have ID values to store in the DBRef in the database. That's what the error is actually saying.

I don't understand, with the operations OP1 and OP3 I'm creating all objects, meanwhile with OP4 I'm updating an object. Now I don't remember exactly what I had in mongodb after OP4, because I removed the reference of the Execution object inside the Message in order to make it work, but still I don't understand which is the wrong workflow.

Cheers.

**Comment by evanchooly** [comment]
Morphia's complaining that it can't fetch that reference because the other document doesn't exist (yet?) in the database.  That validation was removed in 0.111 (iirc) and so you won't see *that* error but you'll still get a failure when you try to use that referenced object.  Can you try with the latest and confirm for me?

In the meantime, I'll try to get your example code working here and see what's happening.

**Comment by evanchooly** [comment]
OK.  I understand, more or less, the problem here.  References don't quite work with compound IDs just yet.  There are some encoding/decoding issues that will take a bit to untangle.  Unfortunately, I doubt the fix will make 1.0 at this point but once we have a stable 1.0, I'll make an attempt to straighten all this out.

**Comment by H4ml3t** [comment]
Thank you very much for your work @evanchooly !
